shared/pubsub.js

```diff
@@ -0,0 +1,386 @@
+  [key: string]: JSONType,
+  +type: 'sub',
+  +contractID: string
+}
+
+export type UnsubMessage = {
+  [key: string]: JSONType,
+  +type: 'unsub',
+  +contractID: string
+}
+
+// ====== Enums ====== //
+
+export const NOTIFICATION_TYPE = Object.freeze({
+  ENTRY: 'entry',
+  PUB: 'pub',
+  SUB: 'sub',
+  UNSUB: 'unsub'
+})
+
+export const REQUEST_TYPE = Object.freeze({
+  PUB: 'pub',
+  SUB: 'sub',
+  UNSUB: 'unsub'
+})
+
+export const RESPONSE_TYPE = Object.freeze({
+  ERROR: 'error',
+  SUCCESS: 'success'
+})
+
+// ====== API ====== //
+
+/**
+ * Creates a pubsub client instance.
+ *
+ * @param {string} url - A WebSocket URL to connect to.
+ * @param {Object?} options
+ * {number?} timeout - Connection timeout duration in milliseconds.
+ * {boolean?} debug
+ * {object?} handlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {boolean?} reconnectOnDisconnection - Whether to reconnect after a server-side disconnection.
+ * {boolean?} reconnectOnOnline - Whether to reconnect after regaining internet connectivity.
+ * {boolean?} reconnectOnTimeout - Whether to reconnect after a timeout.
+ * @returns {Object}
+ */
+export function createClient (url: string, options?: Object = {}): PubSubClient {
+  const client = {
+    customHandlers: options.handlers || {},
+    failedAttempts: 0,
+    isReconnecting: false,
+    listeners: Object.create(null),
+    messageHandlers: { ...defaultMessageHandlers, ...options.messageHandlers },
+    options: { ...defaultOptions, ...options },
+    pendingSubscriptionSet: new Set(),
+    pendingUnsubscriptionSet: new Set(),
+    reconnectionDelayID: undefined,
+    socket: null,
+    subscriptionSet: new Set(),
+    connectionTimeoutID: undefined,
+    url: url.replace(/^http/, 'ws'),
+    // Methods
+    ...{ connect, destroy, getNextRandomReconnectionDelay, pub, sub, unsub }
+  }
+  // Create and save references to reusable event listeners.
+  for (const name of [...eventNames, ...globalEventNames]) {
+    client.listeners[name] = (event: Event, ...args: any[]) => {
+      if (client.options.debug) {
+        console.debug('[pubsub] Event:', name, ...args)
+      }
+      const customHandler = client.customHandlers['on' + name]
+      const defaultHandler = defaultHandlers['on' + name]
+
+      if (defaultHandler) {
+        defaultHandler.call(client, event)
+      }
+      if (customHandler) {
+        customHandler.call(client, event)
+      }
+    }
+  }
+  // Attach global event listeners before the first connection.
+  if (typeof window === 'object') {
+    globalEventNames.forEach((name) => {
+      window.addEventListener(name, client.listeners[name])
+    })
+  }
+  client.connect()
+  return client
+}
+
+export function createMessage (type: string, data: JSONType): string {
+  return JSON.stringify({ type, data })
+}
+
+export function createRequest (type: $Values<typeof REQUEST_TYPE>, data: JSONObject): string {
```

Comment:
I love this `$Values<typeof REQUEST_TYPE>` stuff. You are really a Flowtype master! 👍

Reply:
Thank you. I hope this is not too verbose!😄

Reply:
Well, it is quite verbose, but also cool that JavaScript with Flow can do this kind of stuff if it's necessary. As long as it's sparingly used and not all over the place as it happens with C++, it's fine. :)